### PR TITLE
Reworked module building system

### LINKED ISSUE #17: Shouldn't export UMD build as "main"
`bundle.js` is currently exported as "main". Since this is already bundled, if you import this module with browserify or webpack, it won't be able to trace dependencies correctly. This means you could end up with duplicate dependencies, etc.

Ideally, main should export code that is transpiled, but not bundled. This ensures that exported code is backwards compatible.

A typical node module setup is structured like the following:

src <- Raw source
lib <- Transpiled source
dist <- Bundle distribution

So if we do have a lib directory, the code from lib is what should be exported.

### LINKED ISSUE #18: All of Lodash is built into bundle when running Browserify
Since browserify doesn't use tree shaking by default, all of lodash appears to be bundled into the UMD build by default, which greatly increases the size of this module. We should replace our lodash usage with plain old javascript.

### LINKED ISSUE #23: Don't require manually rebuilding bundle.js before publishing
By adding the following, we should be able to ensure that it is always rebuild before publishing to npm.

```
"scripts": {
    ...
    "prepare": "npm run build"
  }
```

https://docs.npmjs.com/misc/scripts

This should run `npm run build` in the following two instances:
- Before publishing, so when installing the module from `npm` as a dependency, the built bundle AND the lib resources will be available.
- After cloning the project and running `npm install`, which is helpful if someone clones this repository locally and requires it as a module.

This will also mean that `bundle.js` should now be removed from the repo and .gitignored

NOTE: This will also fix an issue we currently have, which is that the `module` export in package.json reference a file in `dist`, but `dist` is never created when a package is installed, so it's completely invalid.

### LINKED ISSUE #24: Rework `dist`, `lib`, and `public` directories
The current `dist` directory should be renamed to `lib`, since it currently holds the transpiled source. 

A NEW `dist` directory should be created to hold bundle.js, which is our UMD distribution.

This is to keep in-line with the following, which is semi-standard for NPM packages: https://stackoverflow.com/questions/39553079/difference-between-lib-and-dist-folders-when-packaging-library-using-webpack

src <- Raw source
lib <- Transpiled source
dist <- Bundled distribution or UMD build

The `lib` dir is what should then be exported from `main` / `browser` / `module` in package.json. 

This is what most packaging systems will be expecting (webpack, browserify, etc.), as opposed to built bundle, which is what is currently exported from `main`.

### PR DESCRIPTION
**Pull request by JasonStoltz**
[ENG-1074] Iron out js client packaging support

### Added Prettier + Huskey

Reformatted all of the Javascript in this project to be formatted with Prettier. Also added Husky which will helps run a precommit hook which runs prettier on all Javascript before it is committed. This is just a simple change I made while I was in there to tidy things up.

### Replaced Webpack with Rollup for build system

Generally speaking, the reason I switched were:
- Rollup, very generally, seems to have be much simpler to configure
- Rollup examples are very much geared specifically towards publishing libraries, which makes them much more relevant to this project and much easier to find answers and examples for how to do the things I was trying to accomplish in this PR.

### Fixes #24: Rework `dist`, `lib`, and `public` directories 

dist <- Now contains bundled distributions of various flavors

  - swiftype_app_search.es.js < ES Modules Build, which Rollup wants
  - swiftype_app_search.umd.js < UMD Build, which browsers, browserify, and webpack want

src <- Raw source
public <- Deleted, using dist instead
lib <- Deleted, not keeping transpiled source around

### Fixes #23: Don't require manually rebuilding bundle.js before publishing

The fix was to add `prepare: "npm run build"` to package.json. This triggers the build just before the project is published, and when `npm install` is run. So if the project is added as a dependency, the built code will be there, because it's part of the published distribution. Additionally, if a user were to check the repository our and run `npm install` and add it locally, the built code will be there as well.

### Fixes #17: Shouldn't export UMD build as "main"

As it turns out, I am still exporting the UMD build as "main". The following is the approach we're taking, in package.json:

`browser` <- Exports our UMD build. This is what Webpack and Browserify look for by default, over any other property (`main` or `module`). Thus it's as if a UMD build is actually what those build tools prefer. Additionally, since our library requires a browser to run (`fetch`, `Header`, etc.), by exporting it using the `browser` key, it clues in potential users of this library that it won't run in node, for example.
`main` <- Also exports our UMD build. If you include this module with node, it uses the `main` export and ignores `module` and `browser`. Node / CommonJS can use UMD builds, so this is an appropriate format to export for Node. Using in Node, however, requires polyfilling the browser internals, like `fetch` and `Header`. We could actually not export anything in `main`, to make it especially clear that this isn't meant for usage in Node, however, since it can be worked around I am leaving it as is.
`module` <- Exports our ES modules build. This is especially important to provide for Rollup users. Rollup expects libraries to expose raw ES module builds.

Note that this was tested with webpack, browserify, and rollup in the following project: https://github.com/swiftype/swiftype-client-workshop/tree/master/sandboxes/swiftype-app-search-javascript

### Fixes #18: All of Lodash is built into bundle when running Browserify

The way we were / are building this module with webpack and rollup *should* use tree-shaking to ensure that all of lodash is not bundled in when you are using just one function from lodash. However, lodash is not currently compatible with tree-shaking, so I simply replaced the single `omit` function I was using with a custom `omit` function.

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -38,44 +41,59 @@ export default class Client {
...
+        });
+    });
   }
 
   _request(path, params) {
     const headers = new Headers({
-      'Authorization': `Bearer ${this.apiKey}`,
-      'Content-Type': 'application/json'
-    })
+      Authorization: `Bearer ${this.apiKey}`,
```
Comment (goodroot) [inline src/client.js]: No quotes required for Authorization?

Comment (JasonStoltz) [reply src/client.js]: Nope, not in JavaScript.

Comment (brianearwood) [reply src/client.js]: ha, I had the same question

Comment (goodroot) [reply src/client.js]: JavaScript is weird. 😆

File: src/client.js [hunk 2]
```diff
@@ -38,44 +41,59 @@ export default class Client {
...
+      "Content-Type": "application/json"
```
Comment (brianearwood) [inline src/client.js]: Why does this one need quotes?

Comment (brianearwood) [reply src/client.js]: Because of the dash, I'm assuming...

Comment (JasonStoltz) [reply src/client.js]: Correct.

File: src/client.js (hunk 1 above)
Comment (JasonStoltz) [reply src/client.js]: Gross would also be an acceptable answer.

**Pull request merged**